File: translated/f-to-c/DRB035-truedepscalar-orig-yes.c
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.

SPDX-License-Identifier: (BSD-3-Clause)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//Loop carried true dep between tmp =..  and ..= tmp.
//Data race pair: tmp@24:9:W vs. tmp@25:15:R

#include <omp.h>
#include <stdio.h>
#include <stdlib.h>

int main()
{
    int i, tmp, len;
    int* a;

    len = 100;
    tmp = 10;
    a = (int*)malloc(len * sizeof(int));

    #pragma omp parallel for
    for (i = 0; i < len; i++) {
        a[i] = tmp;
        tmp = a[i] + i + 1;  // Adjust for 0-based indexing
    }

    printf("a(50) = %d\n", a[49]);

    free(a);

    return 0;
} 

File: translated/f-to-go-automodel-bad/DRB035-truedepscalar-orig-yes.go
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
// and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.
//
// SPDX-License-Identifier: (BSD-3-Clause)
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

//Loop carried true dep between tmp =..  and ..= tmp.
//Data race pair: tmp@24:9:W vs. tmp@25:15:R

package main

import (
	"fmt"
	"runtime"
	"sync"
)

func main() {
	var tmp, length int
	var a []int

	length = 100
	tmp = 10
	a = make([]int, length)

	//$omp parallel do
	var wg sync.WaitGroup
	numCPU := runtime.NumCPU()
	chunkSize := length / numCPU
	if chunkSize < 1 {
		chunkSize = 1
	}

	for start := 1; start <= length; start += chunkSize {
		end := start + chunkSize - 1
		if end > length {
			end = length
		}
		wg.Add(1)
		go func(start, end int) {
			defer wg.Done()
			for i := start; i <= end; i++ {
				a[i-1] = tmp
				tmp = a[i-1] + i
			}
		}(start, end)
	}
	wg.Wait()
	//$omp end parallel do

	fmt.Printf("a(50) =%3d\n", a[49])
}
